data/transactions/logic/resources_test.go

```diff
@@ -0,0 +1,743 @@
+		AssetSender:   otherAcct,
+	}
+	afrz := txntest.Txn{
+		Type:          protocol.AssetFreezeTx,
+		FreezeAsset:   200, // must be < 256, later code assumes it fits in a byte
+		Sender:        senderAcct,
+		FreezeAccount: otherAcct,
+	}
+
+	sources := []string{"", senderBalance}
+	rsources := []string{senderBalance, ""}
+	for _, send := range []txntest.Txn{keyreg, pay, acfg, axfer, afrz} {
+		logic.TestApps(t, sources, txntest.Group(&send, &appl), 9, ledger)
+		logic.TestApps(t, rsources, txntest.Group(&appl, &send), 9, ledger)
+
+		logic.TestApps(t, sources, txntest.Group(&send, &appl), 8, ledger,
+			logic.NewExpect(1, "invalid Account reference"))
+		logic.TestApps(t, rsources, txntest.Group(&appl, &send), 8, ledger,
+			logic.NewExpect(0, "invalid Account reference"))
+	}
+
+	holdingAccess := `
+	txn ApplicationArgs 0
+	txn ApplicationArgs 1; btoi
+	asset_holding_get AssetBalance
+	pop; pop; int 1
+`
+	sources = []string{"", holdingAccess}
+	rsources = []string{holdingAccess, ""}
+
+	t.Run("keyreg", func(t *testing.T) {
+		appl.ApplicationArgs = [][]byte{senderAcct[:], {200}}
+		logic.TestApps(t, sources, txntest.Group(&keyreg, &appl), 9, ledger,
+			logic.NewExpect(1, "invalid Asset reference 200"))
+		withRef := appl
+		withRef.ForeignAssets = []basics.AssetIndex{200}
+		logic.TestApps(t, sources, txntest.Group(&keyreg, &withRef), 9, ledger,
+			logic.NewExpect(1, "invalid Holding access "+senderAcct.String()))
+	})
+	t.Run("pay", func(t *testing.T) {
+		// The receiver is available for algo balance reading
+		appl.ApplicationArgs = [][]byte{receiverAcct[:]}
+		logic.TestApps(t, []string{"", receiverBalance}, txntest.Group(&pay, &appl), 9, ledger)
+
+		// The other account is not (it's not even in the pay txn)
+		appl.ApplicationArgs = [][]byte{otherAcct[:]}
+		logic.TestApps(t, []string{"", otherBalance}, txntest.Group(&pay, &appl), 9, ledger,
+			logic.NewExpect(1, "invalid Account reference "+otherAcct.String()))
+
+		// The other account becomes accessible because used in CloseRemainderTo
+		withClose := pay
+		withClose.CloseRemainderTo = otherAcct
+		logic.TestApps(t, []string{"", otherBalance}, txntest.Group(&withClose, &appl), 9, ledger)
+	})
+
+	t.Run("acfg", func(t *testing.T) {
+		// The other account is not available even though it's all the extra addresses
+		appl.ApplicationArgs = [][]byte{otherAcct[:]}
+		logic.TestApps(t, []string{"", otherBalance}, txntest.Group(&acfg, &appl), 9, ledger,
+			logic.NewExpect(1, "invalid Account reference "+otherAcct.String()))
+	})
+
+	t.Run("axfer", func(t *testing.T) {
+		// The receiver is also available for algo balance reading
+		appl.ApplicationArgs = [][]byte{receiverAcct[:]}
+		logic.TestApps(t, []string{"", receiverBalance}, txntest.Group(&axfer, &appl), 9, ledger)
+
+		// as is the "other" (AssetSender)
+		appl.ApplicationArgs = [][]byte{otherAcct[:]}
+		logic.TestApps(t, []string{"", otherBalance}, txntest.Group(&axfer, &appl), 9, ledger)
+
+		// receiver holding is available
```

Comment:
Could also test other holding is available here

Reply:
Done, and some more related tests (afrz holdings, for example).